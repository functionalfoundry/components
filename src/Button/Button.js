/* @flow */
import React from 'react'
import Theme from 'js-theme'
import mergeProps from 'js-theme/lib/mergeProps'
import {
  Borders,
  Colors,
  Corners,
  Fonts,
  Spacing,
} from '@workflo/styles'

// TODO: Figure out vertical alignment problem ðŸ˜¡

type PropsT = {
  children: React.Children,
  kind: 'regular' | 'secondary' | 'hero',
  shade: 'light' | 'dark',
  disabled: boolean,
  label: string,
  ghost: boolean,
  round: boolean,
  theme: Object,
}

const defaultProps = {
  kind: 'regular',
  shade: 'dark',
  theme: {},
}

const Button = ({
  children,
  label,
  theme,
  disabled, // eslint-disable-line no-unused-vars
  kind, // eslint-disable-line no-unused-vars
  shade, // eslint-disable-line no-unused-vars
  round, // eslint-disable-line no-unused-vars
  ghost, // eslint-disable-line no-unused-vars
  ...props,
}: PropsT) => (
  <button
    {...mergeProps(theme.button, props)}
  >
    {children}
    {label}
  </button>
)

Button.defaultProps = defaultProps

const defaultTheme = ({
  kind,
  ghost,
  // round,
  shade,
}: PropsT) => ({
  button: {
    ...Fonts.base,
    ...Borders.light,
    ...Corners.round,
    whiteSpace: 'nowrap',
    userSelect: 'none',
    textDecoration: 'none',
    boxSizing: 'border-box',
    cursor: 'pointer',
    display: 'flex',
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    border: 0,
    padding: `${Spacing.tiny}px ${Spacing.base}px`,
    textTransform: 'uppercase',
    ':focus': {
      outline: 'none',
    },
    ...getKindStyle(kind, shade, ghost),
    // ...getRoundStyle(round),
  },
})

const getKindStyle = (kind: string, shade: string, ghost: boolean) => {
  switch (`${shade}:${kind}${ghost ? ':ghost' : ''}`) {
    case 'dark:regular':
      return {
        color: Colors.grey200,
        backgroundColor: Colors.grey800,
        border: 0,
        ':hover': {
          backgroundColor: Colors.grey700,
          color: Colors.primary,
        },
      }
    case 'dark:regular:ghost':
      return {
        color: Colors.grey200,
        backgroundColor: 'rgba(0,0,0,0)',
        border: `1px solid ${Colors.grey700}`,
        ':hover': {
          border: `1px solid ${Colors.primary}`,
          color: Colors.primary,
        },
      }
    case 'dark:secondary':
      return {
        color: Colors.primary,
        backgroundColor: Colors.grey900,
        ':hover': {
          backgroundColor: Colors.grey800,
          color: 'white',
        },
      }
    case 'dark:secondary:ghost':
      return {
        color: 'white',
        backgroundColor: 'rgba(0,0,0,0)',
        border: '1px solid rgba(0,0,0,0)',
        ':hover': {
          border: `1px solid ${Colors.primary}`,
        },
      }
    case 'dark:hero':
      return {
        color: Colors.grey900,
        backgroundColor: Colors.primary,
        ':hover': {
          backgroundColor: Colors.primary600,
        },
      }
    case 'dark:hero:ghost':
      return {
        color: Colors.grey200,
        backgroundColor: 'rgba(0,0,0,0)',
        border: `1px solid ${Colors.primary}`,
        ':hover': {
          border: `1px solid ${Colors.primary700}`,
        },
      }
    default: {
      return {}
    }
  }
}

export default Theme('Button', defaultTheme)(Button)
