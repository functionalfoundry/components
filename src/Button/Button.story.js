import React from 'react'
import { storiesOf } from '@kadira/storybook'
import Button from '.'
import Preview from '../Preview'
import PreviewContainer from '../Preview/PreviewContainer'

storiesOf('Button', module)
  .add('Regular', () => (
    <PreviewContainer
      shade='dark'
    >
      <Preview
        label='Regular'
      >
        <Button
          label='Click Me'
        />
      </Preview>
      <Preview
        label='Regular Ghost'
      >
        <Button
          label='Click Me'
          ghost
        />
      </Preview>
      <Preview
        label='Secondary'
      >
        <Button
          label='Click Me'
          kind='secondary'
        />
      </Preview>
      <Preview
        label='Secondary Ghost'
      >
        <Button
          label='Click Me'
          kind='secondary'
          ghost
        />
      </Preview>
      {/* <Preview
        label='Hero'
      >
        <Button
          label='Click Me'
          kind='hero'
        />
      </Preview> */}
      <Preview
        label='Hero Ghost'
      >
        <Button
          label='Click Me'
          kind='hero'
          ghost
        />
      </Preview>
    </PreviewContainer>
  ))
