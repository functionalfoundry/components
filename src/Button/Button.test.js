import React from 'react'
import renderer from 'react-test-renderer'
import Button from '.'

test('Renders without props', () => {
  const button = renderer.create(
    <Button />
  ).toJSON()
  expect(button).toMatchSnapshot()
})

test('Renders regular ghost on dark', () => {
  const button = renderer.create(
    <Button
      shade='dark'
      ghost
    />
  ).toJSON()
  expect(button).toMatchSnapshot()
})

test('Renders unknown button kinds', () => {
  const button = renderer.create(
    <Button
      kind='unknown'
    />
  ).toJSON()
  expect(button).toMatchSnapshot()
})
