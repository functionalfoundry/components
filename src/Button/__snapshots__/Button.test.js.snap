exports[`test Renders regular ghost on dark 1`] = `
<button
  className="button_bxfwpx" />
`;

exports[`test Renders unknown button kinds 1`] = `
<button
  className="button_lyx963" />
`;

exports[`test Renders without props 1`] = `
<button
  className="button_qji6qz" />
`;
